File: Flee/ShipStats.vb
Imports System.Globalization

Public Class ShipStats

	' shared
	Public Shared classes As New Dictionary(Of String, ShipStats)
	Public Shared Function DumpClasses() As String
		Dim total As String = ""
		For Each stats As ShipStats In classes.Values
			total += stats.ToString()
			total += vbLf
		Next
		Return total
	End Function

	' identity
	Public name As String = "Default"
	Public desc As String = Nothing

	' visual
	Public sprite As String = "Plasma"
	Public level As Integer = 0
	Public width As Integer = -1

	' stats (base)
	Public integrity As Integer = 0
	Public repair As Integer = 0
	Public speed As Double = 0.0
	Public turn As Double = 0.0
	Public default_weapons As New List(Of String)

	' stats (shields)
	Public shield As Integer = 0
	Public shield_regeneration As Integer = 10
	Public shield_opacity As Double = 0

	' stats (deflectors)
	Public deflectors As Integer = 0
	Public deflectors_cooldown As Integer = 64
	Public hot_deflector As Double = 0
	Public cold_deflector As Boolean = 0

	' crafting
	Public crafts As New List(Of String)
	Public cost As MaterialSet = New MaterialSet(0, 0, 0, 0)
	Public complexity = 0

	' upgrades
	Public native_upgrades As New List(Of String)

	' constructor
	Public Sub New(name As String)
		Me.name = name
		SetSprite(name)
	End Sub
	Public Sub SetSprite(sprite As String)
		Me.sprite = sprite
		Dim try_bmp As Bitmap = Helpers.GetSprite(Me.sprite, -1, -1, Nothing)
		If Not try_bmp Is Nothing Then
			Me.width = try_bmp.Width / 8 - 2
			Me.level = Math.Sqrt(Me.width)
			Me.integrity = (Me.width * Me.width) / 12
			If Me.width >= 25 Then Me.repair = 1
			Me.cost.Metal = Me.width * 2 + Me.level * 10
			If Me.width >= 20 Then Me.cost.Crystal = 1
		End If
	End Sub

	' Import/Export
	Public Sub SetProperty(name As String, value As String)
		Select Case name
			Case "desc" : Me.desc = value
			Case "sprite"
				SetSprite(value)
			Case "level" : Me.level = Convert.ToInt32(value)
			Case "width" : Me.width = Convert.ToInt32(value)
			Case "integrity" : Me.integrity = Convert.ToInt32(value)
			Case "repair" : Me.repair = Convert.ToInt32(value)
			Case "speed"
				Me.speed = Helpers.ToDouble(value)
				If turn = 0 Then Me.turn = speed
			Case "turn" : Me.turn = Helpers.ToDouble(value)
			Case "weapon" : default_weapons.Add(value)
			Case "shield"
				Me.shield = value
				If Me.shield_opacity = 0 Then Me.shield_opacity = 25
			Case "shield_regeneration" : Me.shield_regeneration = Helpers.ToDouble(value)
			Case "shield_opacity" : Me.shield_opacity = Helpers.ToDouble(value)
			Case "deflectors" : Me.deflectors = Convert.ToInt32(value)
			Case "deflectors_cooldown" : Me.deflectors_cooldown = Convert.ToInt32(value)
			Case "hot_deflector" : Me.hot_deflector = Helpers.ToDouble(value)
			Case "cold_deflector" : Me.cold_deflector = Convert.ToInt32(value)
			Case "craft" : Me.crafts.Add(value)
			Case "native_upgrade" : Me.native_upgrades.Add(value)
			Case "cost"
				Me.cost = New MaterialSet(value)
				If Me.complexity = 0 Then Me.complexity = Me.width * 5 + Me.cost.Metal / 8 + Me.cost.Crystal * 15 + Me.cost.Antimatter / 4 + Me.cost.Fissile * 100
			Case "complexity" : Me.complexity = Convert.ToInt32(value)
			Case Else : Throw New Exception("'" & name & "' is not a valid ship property")
		End Select
	End Sub
	Public Function ToString() As String
		Dim total As String = "ship " & Me.name & vbLf
		If Me.sprite <> Me.name Then
			total &= vbTab & "sprite=" & Me.sprite & vbLf
		End If
		total &= vbTab & "level=" & Me.level.ToString() & vbLf
		If Me.width <> Helpers.GetSprite(Me.sprite, -1, -1, Nothing).Width / 8 - 2 Then
			total &= vbTab & "width=" & Me.width.ToString() & vbLf
		End If
		total &= vbTab & "integrity=" & Me.integrity.ToString() & vbLf
		If speed <> 0.0 Then
			total &= vbTab & "speed=" & Helpers.ToString(speed) & vbLf
		End If
		If turn <> speed Then
			total &= vbTab & "turn=" & Helpers.ToString(Me.turn) & vbLf
		End If
		For Each item As String In default_weapons
			total &= vbTab & "weapon=" & item & vbLf
		Next
		'weapons
		If shield > 0 Then
			total &= vbTab & "shield=" & Me.shield.ToString() & vbLf
			If shield_regeneration <> 10 Then
				total &= vbTab & "shield_regeneration=" & Helpers.ToString(Me.shield_regeneration) & vbLf
			End If
			If shield_opacity <> 25 Then
				total &= vbTab & "shield_opacity=" & Helpers.ToString(Me.shield_opacity) & vbLf
			End If
		End If
		If deflectors > 0 Then
			total &= vbTab & "deflectors=" & Me.deflectors.ToString() & vbLf
			If deflectors_cooldown <> 64 Then
				total &= vbTab & "deflectors_cooldown=" & Me.deflectors_cooldown.ToString() & vbLf
			End If
		End If
		If hot_deflector Then
			total &= vbTab & "hot_deflector=" & Helpers.ToString(Me.hot_deflector) & vbLf
		End If
		If cold_deflector Then
			total &= vbTab & "cold_deflector=" & Convert.ToInt32(Me.cold_deflector).ToString() & vbLf
		End If
		For Each item As String In crafts
			total &= vbTab & "craft=" & item & vbLf
		Next
		For Each item As String In native_upgrades
			total &= vbTab & "native_upgrade=" & item & vbLf
		Next
		total &= vbTab & "cost=" & Me.cost.ToString() & vbLf
		total &= vbTab & "complexity=" & Me.complexity.ToString() & vbLf
		Return total
	End Function
	Public Function Clone() As ShipStats
		Return DirectCast(Me.MemberwiseClone(), ShipStats)
	End Function

End Class

File: Flee/ShipStats.cs
﻿using System;
using System.Collections.Generic;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace Flee {

	/**
	 * @brief Special flags representing ships gameplay role.
	 */
	public enum ShipRole {
		Shipyard = 0x1, // Capable of summoning others. It keeps the team status as alive.
		Defense = 0x2, // Summoned with Shipyards.
		Starter = 0x4, // May be summoned with a Shipyard in a player team.
		Playable = 0x10, // May be summoned in a player team.
		NPC = 0x20, // May be summoned in an NPC team.
		Boss = 0x40, // May be summoned as a Boss.
		Derelict = 0x80 ,// May be summoned as a derelict object.
		Static = 0x100, // May be summoned as a static object such as a star.
		Mine = 0x200, // Is mined instead of looted.
	}

	/**
	 * @brief Represent a Ship base stats.
	 */
	public class ShipStats {

		/* role */
		static int ShipRoles(string roles_str) {
			int total = 0;
			foreach (string role_str in roles_str.Split('|')) {
				total |= (int)Enum.Parse(typeof(ShipRole), role_str, true);
			}
			return (total);
		}
		static string ShipRoles(int roles) {
			string total = "";
			foreach (ShipRole role in Enum.GetValues(typeof(ShipStats))) {
				if ((roles & (int)role) != 0) {
					if (total.Length > 0)
						total += '|';
					total += role.ToString();
				}
			}
			return (total);
		}

		// shared
		public static Dictionary<string, ShipStats> classes = new Dictionary<string, ShipStats>();
		public static string DumpClasses() {
			string total = "";
			foreach (ShipStats stats in classes.Values) {
				total += stats.ToString();
				total += Constants.vbLf;
			}

			return total;
		}

		// spawning rules
		public double spawning_frequency = 1.0;
		public int spawning_amount_min = 1;
		public int spawning_amount_max = 2;

		// identity
		public string name = "Default";
		public string desc = null;
		public int role = 0; // ShipRole flags

		// visual
		public string sprite = "Plasma";
		public int level = 0;
		public int width = -1;

		// stats (base)
		public float integrity = 0;
		public float repair = 0;
		public float speed = 0.0f;
		public float turn = 0.0f;
		public List<string> default_weapons = new List<string>();

		// stats (shields)
		public float shield = 0;
		public float shield_regeneration = 0.0005f;
		public float shield_opacity = 0.0f;

		// stats (deflectors)
		public int deflectors = 0;
		public int deflectors_cooldown = 128;
		public int hot_deflectors = 0;
		public int cold_deflectors = 0;
		public int TotalDeflectorsMax() { return (deflectors + hot_deflectors); }

		// crafting
		public List<string> crafts = new List<string>();
		public MaterialSet cost = new MaterialSet(0L, 0L, 0L, 0L);
		public int complexity = 0;

		// upgrades
		public List<string> native_upgrades = new List<string>();

		// constructor
		public ShipStats(string name) {
			this.name = name;
			try {
				SetSprite(name);
			} catch (Exception) {
			}
		}

		public void SetSprite(string sprite) {
			this.sprite = sprite;
			var try_bmp =  SpriteArray.GetSpriteArray(this.sprite).GetSprite(0, 0);
			if (try_bmp is object) {
				width = try_bmp.Width;
				level = (int)Math.Sqrt(width) / 2;
				integrity = (int)(width * width / 12d);
				if (width >= 25)
					repair = 1;
				cost.Metal = width * 2 + level * 10;
				if (width >= 20)
					cost.Crystal = 1L;
			}
		}

		// Import/Export
		public void SetProperty(string name, string value) {
			switch (name ?? "") {
			case "desc": {
				desc = value;
				break;
			}
			case "sprite": {
				SetSprite(value);
				break;
			}
			case "level": {
				level = Convert.ToInt32(value);
				break;
			}
			case "width": {
				width = Convert.ToInt32(value);
				break;
			}
			case "integrity": {
				integrity = Convert.ToInt32(value);
				break;
			}
			case "repair": {
				repair = Convert.ToInt32(value);
				break;
			}
			case "speed": {
				speed = Helpers.ToFloat(value);
				if (turn == 0d)
					turn = speed;
				break;
			}
			case "turn": {
				turn = Helpers.ToFloat(value);
				break;
			}
			case "weapon": {
				default_weapons.Add(value);
				break;
			}
			case "shield": {
				shield = Convert.ToInt32(value);
				if (shield_opacity == 0.0)
					shield_opacity = 0.25f;
				break;
			}
			case "shield_regeneration": {
				shield_regeneration = Helpers.ToFloat(value);
				break;
			}
			case "shield_opacity": {
				shield_opacity = Helpers.ToFloat(value);
				break;
			}
			case "deflectors": {
				deflectors = Convert.ToInt32(value);
				break;
			}
			case "deflectors_cooldown": {
				deflectors_cooldown = Convert.ToInt32(value);
				break;
			}
			case "hot_deflectors": {
				hot_deflectors = Convert.ToInt32(value);
				break;
			}
			case "cold_deflectors": {
				cold_deflectors = Convert.ToInt32(value);
				break;
			}
			case "craft": {
				crafts.Add(value);
				break;
			}
			case "native_upgrade": {
				native_upgrades.Add(value);
				break;
			}
			case "cost": {
				cost = new MaterialSet(value);
				if (complexity == 0)
					complexity = (int)((width * 5) + (cost.Metal / 8) + (cost.Crystal * 15L) + (cost.Starfuel / 4d) + (cost.Fissile * 100L));
				break;
			}
			case "complexity": {
				complexity = Convert.ToInt32(value);
				break;
			}
			// spawning
			case "role": role = ShipRoles(value); break;
			case "spawning_frequency": spawning_frequency = Helpers.ToDouble(value); break;
			case "spawning_amount_min": spawning_amount_min = Convert.ToInt32(value); break;
			case "spawning_amount_max": spawning_amount_max = Convert.ToInt32(value); break;
			case "spawning_amount": {
				spawning_amount_min = Convert.ToInt32(value);
				spawning_amount_max = spawning_amount_min + 1;
				break;
			}
			default: throw new Exception("'" + name + "' is not a valid ship property");
			}
		}

		public override string ToString() {
			string total = "ship " + name + Constants.vbLf;
			if ((sprite ?? "") != (name ?? ""))
				total += Constants.vbTab + "sprite=" + sprite + Constants.vbLf;

			total += Constants.vbTab + "level=" + level.ToString() + Constants.vbLf;
			if (width != Helpers.GetSprite(sprite, -1, -1, default).Width / 8d - 2d)
				total += Constants.vbTab + "width=" + width.ToString() + Constants.vbLf;

			total += Constants.vbTab + "integrity=" + integrity.ToString() + Constants.vbLf;
			if (speed != 0.0d)
				total += Constants.vbTab + "speed=" + Helpers.ToString(speed) + Constants.vbLf;

			if (turn != speed)
				total += Constants.vbTab + "turn=" + Helpers.ToString(turn) + Constants.vbLf;

			foreach (string item in default_weapons)
				total += Constants.vbTab + "weapon=" + item + Constants.vbLf;
			// weapons
			if (shield > 0) {
				total += Constants.vbTab + "shield=" + shield.ToString() + Constants.vbLf;
				if (shield_regeneration != 10)
					total += Constants.vbTab + "shield_regeneration=" + Helpers.ToString(shield_regeneration) + Constants.vbLf;

				if (shield_opacity != 25d)
					total += Constants.vbTab + "shield_opacity=" + Helpers.ToString(shield_opacity) + Constants.vbLf;
			}

			if (deflectors > 0) {
				total += Constants.vbTab + "deflectors=" + deflectors.ToString() + Constants.vbLf;
				if (deflectors_cooldown != 64)
					total += Constants.vbTab + "deflectors_cooldown=" + deflectors_cooldown.ToString() + Constants.vbLf;
			}

			if (hot_deflectors != 0)
				total += "\thot_deflectors=" + hot_deflectors.ToString() + "\n";

			if (cold_deflectors != 0)
				total += "\tcold_deflectors=" + cold_deflectors.ToString() + "\n";

			foreach (string item in crafts)
				total += Constants.vbTab + "craft=" + item + Constants.vbLf;
			foreach (string item in native_upgrades)
				total += Constants.vbTab + "native_upgrade=" + item + Constants.vbLf;
			total += Constants.vbTab + "cost=" + cost.ToString() + Constants.vbLf;
			total += Constants.vbTab + "complexity=" + complexity.ToString() + Constants.vbLf;
			// spawning
			if (role != 0)
				total += "role=" + ShipRoles(role) + "\n";
			if (spawning_frequency != 1.0f)
				total += "\tspawning_frequency=" + Helpers.ToString(spawning_frequency) + "\n";
			if (spawning_amount_min != 1)
				total += "\tspawning_amount_min=" + spawning_amount_min.ToString() + "\n";
			if (spawning_amount_max != 2)
				total += "\tspawning_amount_max=" + spawning_amount_max.ToString() + "\n";
			return total;
		}

		public ShipStats Clone() {
			return (ShipStats)MemberwiseClone();
		}
	}
}